ctapipe/tools/fit_spe.py

```diff
@@ -0,0 +1,225 @@
+
+
+def append_to_table(pixel, table, result):
+    row = table.row
+    for key, value in result[pixel].items():
+        row[key] = value
+    row.append()
+
+
+def append_illumination_to_table(illumination, pixel, table, result):
+    row = table.row
+    for key, value in result[pixel][illumination].items():
+        row[key] = value
+    row.append()
+
+
+def write_spe_file(fitter, pdf, n_pixels, metadata, output_path):
+    hist_size = fitter.pixel_arrays[0][0]['charge_hist_x'].size
+    curve_size = fitter.pixel_arrays[0][0]['fit_x'].size
+
+    # Prepare Table descriptions
+    class ParamTableDesc(tables.IsDescription):
+        pass
+    for parameter in pdf.parameter_names:
+        ParamTableDesc.columns[parameter] = tables.Float64Col()
+
+    class ScoreTableDesc(tables.IsDescription):
+        chi2 = tables.Float64Col()
+        reduced_chi2 = tables.Float64Col()
+        p_value = tables.Float64Col()
+
+    class ArrayTableDesc(tables.IsDescription):
+        charge_hist_x = tables.Float64Col(shape=(hist_size))
+        charge_hist_y = tables.Float64Col(shape=(hist_size))
+        charge_hist_edges = tables.Float64Col(shape=(hist_size+1))
+        fit_x = tables.Float64Col(shape=(curve_size))
+        fit_y = tables.Float64Col(shape=(curve_size))
+
+    Provenance().add_output_file(output_path)
+    with tables.File(output_path, mode='w') as output:
+        results = output.create_group(output.root, "results", "results")
+        arrays = output.create_group(output.root, "arrays", "arrays")
+        values_table = output.create_table(
+            results, "values", ParamTableDesc, "SPE Fit Parameter Values"
+        )
+        errors_table = output.create_table(
+            results, "errors", ParamTableDesc, "SPE Fit Parameter Errors"
+        )
+        scores_table = output.create_table(
+            results, "scores", ScoreTableDesc, "SPE Fit Scores"
+        )
+
+        for pixel in range(n_pixels):
+            append_to_table(pixel, values_table, fitter.pixel_values)
+            append_to_table(pixel, errors_table, fitter.pixel_errors)
+            append_to_table(pixel, scores_table, fitter.pixel_scores)
+
+        for i in range(fitter.n_illuminations):
+            arrays_table = output.create_table(
+                arrays,
+                f"illumination_{i}",
+                ArrayTableDesc,
+                f"SPE Fit Illumination {i} Arrays"
+            )
+            for pixel in range(n_pixels):
+                append_illumination_to_table(i, pixel, arrays_table, fitter.pixel_arrays)
+
+        metadata_node = filenode.new_node(output, where='/', name='metadata')
+        for key, value in metadata.items():
+            metadata_node.attrs[key] = value
+
+
+class SPEFitter(Tool):
+    name = "SPEFitter"
+    description = "Extract the Single Photoelectron spectrum parameters from the measured charges stored in a DL1 file"
+
+    input_paths = List(Unicode(), None, help="Paths to DL1 files").tag(config=True)
```

Comment:
Also use paths here.

Note, that you cannot give these on the command line.

I would propose to use an `argparse.ArgumentParser` on `self.extra_args` like here with the dl1 merge tool: #1447

Reply:
Yes you can, for example:
```
python fit_spe.py --SPEFitter.input_paths="['/Users/Jason/Downloads/tempdata/spefit_ctapipe/run43516_dl1.h5', '/Users/Jason/Downloads/tempdata/spefit_ctapipe/run43512_dl1.h5']" --SPEFitter.telescope=1 --SPEFitter.pdf_name=SiPMGentile --SPEFitter.output_path=/Users/Jason/Downloads/tempdata/spefit_ctapipe/spe.h5 --SPEFitter.initial="{'pe': 0.6, 'lambda_0': 0.6, 'lambda_1': 1.7, 'opct': 0.4}" --SPEFitter.range="(-1, 8)"
```

Reply:
Also, I have just seen these deprecation messages:
```
/Users/Jason/opt/anaconda3/envs/cta/lib/python3.7/site-packages/traitlets/traitlets.py:2527: FutureWarning: --input_paths=['/Users/Jason/Downloads/tempdata/spefit_ctapipe/run43516_dl1.h5', '/Users/Jason/Downloads/tempdata/spefit_ctapipe/run43512_dl1.h5'] for containers is deprecated in traitlets 5.0. You can pass --input_paths item ... multiple times to add items to a list.
  FutureWarning
/Users/Jason/opt/anaconda3/envs/cta/lib/python3.7/site-packages/traitlets/traitlets.py:2945: FutureWarning: --initial={'pe': 0.6, 'lambda_0': 0.6, 'lambda_1': 1.7, 'opct': 0.4} for dict-traits is deprecated in traitlets 5.0. You can pass --initial <key=value> ... multiple times to add items to a dict.
  FutureWarning,
/Users/Jason/opt/anaconda3/envs/cta/lib/python3.7/site-packages/traitlets/traitlets.py:2527: FutureWarning: --range=(-1, 8) for containers is deprecated in traitlets 5.0. You can pass --range item ... multiple times to add items to a list.
```
So it should be unnecessary to do ones own argparse handling

Reply:
I think the idea was not to parse all arguments, but use ArgParse *after* traitlets has parsed the arguments, in order to allow positional arguments (e.g. a list of input files) as in the merge tool.  That makes it possible to use the shell to expand things as well, like `ctapipe-spefit --range="(-1,8") *.dl1.h5`